Micrometer integration (#17078)
newrelic/docs-website: src/content/docs/more-integrations/open-source-telemetry-integrations/micrometer/micrometer-metrics-registry.mdx

---
title: Micrometer metrics registry
tags:
  - Integrations
  - Open source telemetry integrations
  - Micrometer
title: Forward Micrometer data to New Relic with OpenTelemetry
metaDescription: New Relic offers an integration that sends your Micrometer telemetry data to your New Relic account.
redirects:
  - /docs/integrations/open-source-telemetry-integrations/micrometer/micrometer-metrics-registry
  - /docs/integrations/open-source-telemetry-integrations/open-source-telemetry-integration-list/new-relics-micrometer-integration
    - /docs/integrations/open-source-telemetry-integrations/micrometer/micrometer-metrics-registry
    - /docs/integrations/open-source-telemetry-integrations/open-source-telemetry-integration-list/new-relics-micrometer-integration
freshnessValidatedDate: never
---

> [Micrometer](https://micrometer.io/) provides a simple facade over the instrumentation clients for the most popular monitoring systems, allowing you to instrument your JVM-based application code without vendor lock-in.

Author: Added back!

> You can configure the OpenTelemetry bridge to forward Micrometer data to New Relic. The OpenTelemetry Micrometer bridge directs Micrometer to record telemetry to OpenTelemetry, which you can then export to New Relic with the OpenTelemtry SDK. 

Reviewer: ```suggestion
You can configure the [OpenTelemetry Micrometer Bridge](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/micrometer/micrometer-1.5/library) along with the OpenTelemetry SDK to forward Micrometer data to New Relic via OTLP.
```

> Before you get started, we recommend that review the requirements outlined in the [Getting started](https://github.com/open-telemetry/opentelemetry-java-instrumentation?tab=readme-ov-file#getting-started) section of the OpenTelemetry Java instrumentation repo. Once you've reviewed this section, return to this doc.

Reviewer: The "Getting Started" guide you link to isn't relevant. That's all about using the OpenTelemetry java agent, which performs bytecode manipulation to install instrumentation. We're doing something called library instrumentation, where you add a dependency to a specific instrumentation library, and make code changes to install that instrumentation library in your app. The `opentelemetry-java-instrumentation` repo contains instrumentation which is bundled with the opentelemetry java agent and also stand-alone library instrumentation, so it can be a bit confusing.

Author: Removed! I've updated so this section only contains:

> Before working through these procedures, you should:
> 
> * Configure the OpenTelemetry Metrics SDK to export data to New Relic via [OpenTelemetry Protocol (OTLP)](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-introduction/#how-it-works)
> * Install the OpenTelemetry Micrometer bridge

At minimum, you should:

* Have a Springboot app with Micrometer enabled. Your Springboot app should be instrumented with Micrometer out of the box.

> * Ensure your OpenTelemetry Java agent is the latest version.

Reviewer: - Configure the OpenTelemetry Metrics SDK to export data to New Relic via OTLP
- Install the OpenTelemetry Micrometer bridge

## Forward Micrometer data to New Relic [#forward-data-OpenTelemetry]

<Steps>
    <Step>

### Add OpenTelemetry Micrometer instrumentation

Add OpenTelemetry Micrometer instrumentation to the alpha modules section of your `build.gradle` file:

    ```java
    //Alpha modules
    implementation 'io.opentelemetry.instrumentation:opentelemetry-micrometer-1.5'
    ```

    </Step>
    <Step>

### Add OpenTelemetry API and SDK dependencies
        

> In the `dependencies()` argument, add the OpenTelemetry API and SDK:

Reviewer: ```suggestion
In the `dependencies` section, add the OpenTelemetry SDK and OTLP exporter:
```

    ```java
    dependencies {

>         implementation 'io.opentelemetry:opentelemetry-api'

Reviewer: ```suggestion
```
This isn't really necessary. I'll update the example code to reflect this and other recommendations I'm making here.

>         implementation 'io.opentelemetry:opentelemetry-sdk'

Reviewer: ```suggestion
        implementation 'io.opentelemetry:opentelemetry-sdk'
        implementation 'io.opentelemetry:opentelemetry-exporters-otlp'
```

    }
    ```
    
An example completed file might look something like this:

    ```java
    plugins {
        id 'java-library'
        id 'org.springframework.boot'
    }

    bootRun {
        mainClass.set 'io.opentelemetry.example.micrometer.Application'
    }

    dependencies {
        //OpenTelemetry API and SDK
        implementation 'io.opentelemetry:opentelemetry-api'
        implementation 'io.opentelemetry:opentelemetry-sdk'
        //Alpha modules where Micrometer instrumentation appears
        implementation 'io.opentelemetry.instrumentation:opentelemetry-micrometer-1.5'

>         implementation 'io.opentelemetry:opentelemetry-exporter-otlp'

Reviewer: ```suggestion
```

* Detect Micrometer data
* Forward that data to New Relic

```java
public OpenTelemetry openTelemetry() {
    return OpenTelemetrySdk.builder()
        .setMeterProvider(
            SdkMeterProvider.builder()
                .setResource(
                    Resource.getDefault().toBuilder()
                        .put("service.name", "micrometer-shim")
                        // Include instrumentation.provider=micrometer to enable micrometer metrics
                        // experience in New Relic
                        .put("instrumentation.provider", "micrometer")
                        .build())
                .registerMetricReader(
                    PeriodicMetricReader.builder(
                            OtlpHttpMetricExporter.builder()
                                .setEndpoint("https://otlp.nr-data.net")
                                .addHeader(
                                    "api-key",
                                    Optional.ofNullable(System.getenv("NEW_RELIC_LICENSE_KEY"))
                                        .filter(str -> !str.isEmpty() && !str.isBlank())
                                        .orElseThrow())

>                                 // IMPORTANT: New Relic requires metrics to be delta temporality

Author: What is delta temporality here? What exactly is neW reLIC requiring?

Reviewer: Aggregation temporality determines whether metrics which are reported represent the cumulative or the delta since the last report. For example, if you have a counter which is incremented every minute and you report / export it every minute, cumulative temporality would export values `1, 2, 3, 4, ..`, while delta would export values `1, 1, 1, 1, ...`. 

New Relic is a delta metrics backend. We support cumulative, but for various reasons its better for everyone to send us delta.

Author: woah! TIL! Super cool tidbit, thanks!

                                .setAggregationTemporalitySelector(
                                    AggregationTemporalitySelector.deltaPreferred())
                                // Use exponential histogram aggregation for histogram instruments

>                                 // to

Author: Same as above: what does it mean to 'use histogram aggregation', can we unspool this so the comments are a bit more helpful? (i see that this is the name itself, but i think the comments can elaborate on the name)

Reviewer: When micrometer records measurements against its timer instrument type, e.g. for `http.server.requests`, those are aggregated and exported as a [histogram with explicit bucket boundaries](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#explicit-bucket-histogram-aggregation) by default. Here we're instructing the SDK to use [exponential bucket histogram](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#base2-exponential-bucket-histogram-aggregation) instead, which produces a much more detailed / useful distribution which automatically scales to record the range of measurements recorded to the instruments.

There's obviously a lot to unpack with that, which to do properly would take a few paragraphs. Not sure its worth it to elaborate.

                                // produce better data and compression
                                .setDefaultAggregationSelector(
                                    DefaultAggregationSelector.getDefault()
                                        .with(
                                            InstrumentType.HISTOGRAM,
                                            Aggregation.base2ExponentialBucketHistogram()))
                                .build())
                        // Match default micrometer collection interval of 60 seconds
                        .setInterval(Duration.ofSeconds(60))
                        .build())
                .build())
        .build();
    }
```
    
        
    </Step>

        <Step>

### Find your data in New Relic 

> Wait a few minutes, trigger some test data, then go to **[one.newrelic.com](https://one.newrelic.com) > All Capabilities > APM & Services**, then choose the service instrumented with Micrometer.

Author: is this the correct path to access micrometer-forwarded data?

Reviewer: Better to direct users to "APM & Services -> Services - OpenTelemetry".

If you've instrumented correctly, your data will appear as APM data. 
    </Step>
    
</Steps>

## What's next? [#whats-next]

> To learn more about BLAHBLAHBLAH, see our docs:

Reviewer: Better fix this before merging! 😉

Author: Fixed!

freshnessValidatedDate: never
---

import opentelemetryMicrometerWithOtel from 'images/opentelemetry_screenshot-full_micrometer-with-otel.webp'

New Relic supports [Micrometer data](https://micrometer.io/) so you can view all of your observability metrics in one platform. You'll configure the [OpenTelemetry Micrometer bridge](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/micrometer/micrometer-1.5/library) with the OpenTelemetry SDK, then use OpenTelemetry Protocol (OTLP) to forward your Micrometer data to New Relic.

<img
    title="Micrometer bridged with OpenTelemetry in New Relic"
    alt="A screenshot of the Micrometer summary page when instrumented with OpenTelemetry"
    src={opentelemetryMicrometerWithOtel}
/>

<figcaption>
Go to **[one.newrelic.com](https://one.newrelic.com) > All Capabilities > APM & Services**, then find the **Services - OpenTelemetry** section: View your Micrometer data in New Relic when bridged through OpenTelemetry
</figcaption>

## Compatibility and requirements [#requirements]

> Before working through these procedures, you should:

Reviewer: This isn't quite right. The procedure below tells you how to install the OpenTelemetry micrometer bridge and how to configure the OpenTelemetry metrics SDK to export to New Relic via OTLP. This whole Compatibility and requirements section can probably be deleted, unless you want to talk about  requirement for a license key or something.

<Steps>
    <Step>

### Add OpenTelemetry Micrometer instrumentation

Add OpenTelemetry Micrometer instrumentation to the alpha modules section of your `build.gradle` file:

    ```java
    //Alpha modules
    implementation 'io.opentelemetry.instrumentation:opentelemetry-micrometer-1.5'
    ```

    </Step>
    <Step>

### Add OpenTelemetry dependencies
        
In the `dependencies` section, add the OpenTelemetry SDK and OTLP exporter:

    ```java
    dependencies {
        implementation 'io.opentelemetry:opentelemetry-sdk'
        implementation 'io.opentelemetry:opentelemetry-exporters-otlp'
    }
    ```
    

An example file with alpha modules and dependencies added might look something like this:

```java
plugins {
    id 'java-library'
    id 'org.springframework.boot'
}

bootRun {
    mainClass.set 'io.opentelemetry.example.micrometer.Application'
}

dependencies {
    implementation platform("io.opentelemetry:opentelemetry-bom-alpha:<JAVA_OTEL_VERSION>")
    implementation 'io.opentelemetry:opentelemetry-sdk'
    implementation 'io.opentelemetry:opentelemetry-exporters-otlp'

>     implementation platform("io.opentelemetry.instrumentation:opentelemetry-instrumentation-bom-alpha:<OTEL_JAVA_)INSTRUMENTATION_VERSION>")

Reviewer: Typo. Remove erroneous `)` character.

    implementation 'io.opentelemetry.instrumentation:opentelemetry-micrometer-1.5'
}
```

> Keep in mind you will need to update the snippet with correct versioning. 

Reviewer: How about:
```
Keep in mind you will need to update the snippet to replace `<JAVA_OTEL_VERSION>` and `<OTEL_JAVA_INSTRUMENTATION_VERSION>` with the correct version. See [opentelemetry-java-instrumentation/releases](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases) and [opentelemetry-java/releases](https://github.com/open-telemetry/opentelemetry-java/releases) for more details.
```

    </Step>
    <Step>

### Configure Micrometer to forward data to New Relic

Below is an example code snippet that directs Micrometer to use the OpenTelemetry Micrometer bridge. 

This snippet updates your code so OpenTelemetry can detect Micrometer data, then forward that data to New Relic:

```java
public OpenTelemetry openTelemetry() {
    return OpenTelemetrySdk.builder()
        .setMeterProvider(
            SdkMeterProvider.builder()
                .setResource(
                    Resource.getDefault().toBuilder()
                        .put("service.name", "micrometer-shim")
                        // Include instrumentation.provider=micrometer to enable micrometer metrics
                        // experience in New Relic
                        .put("instrumentation.provider", "micrometer")
                        .build())
                .registerMetricReader(
                    PeriodicMetricReader.builder(
                            OtlpHttpMetricExporter.builder()
                                .setEndpoint("https://otlp.nr-data.net")
                                .addHeader(
                                    "api-key",

>                                     Optional.ofNullable(System.getenv("NEW_RELIC_LICENSE_KEY"))

Reviewer: Should add a note that this depends on you having the `NEW_RELIC_LICENSE_KEY` set to your NR license key. So requires something like:
```
export NEW_RELIC_LICENSE_KEY=<INSERT_LICENSE_KEY>
```

                                // rather than cumulative temporality
                                .setAggregationTemporalitySelector(
                                    AggregationTemporalitySelector.deltaPreferred())
                                // Use exponential histogram aggregation for histogram instruments
                                // to
                                // produce better data and compression
                                .setDefaultAggregationSelector(
                                    DefaultAggregationSelector.getDefault()
                                        .with(
                                            InstrumentType.HISTOGRAM,
                                            Aggregation.base2ExponentialBucketHistogram()))
                                .build())
                        // Match default micrometer collection interval of 60 seconds
                        .setInterval(Duration.ofSeconds(60))
                        .build())
                .build())
        .build();
    }
```
    
        
    </Step>

        <Step>

### Find your data in New Relic 

> Wait a few minutes, trigger some test data, then go to **[one.newrelic.com](https://one.newrelic.com) > All Capabilities > APM & Services**, then find **Services - OpenTelemetry** to choose the service instrumented with Micrometer.

Reviewer: Can also mention that they can find metrics using the [metrics explorer](https://docs.newrelic.com/docs/query-your-data/explore-query-data/browse-data/introduction-data-explorer/).

---
title: Forward Micrometer data to New Relic with OpenTelemetry
metaDescription: New Relic offers an integration that sends your Micrometer telemetry data to your New Relic account.
redirects:
    - /docs/integrations/open-source-telemetry-integrations/micrometer/micrometer-metrics-registry
    - /docs/integrations/open-source-telemetry-integrations/open-source-telemetry-integration-list/new-relics-micrometer-integration
freshnessValidatedDate: never
---

import opentelemetryMicrometerWithOtel from 'images/opentelemetry_screenshot-full_micrometer-with-otel.webp'

> [Micrometer](https://micrometer.io/) provides a simple facade over instrumentation clients, letting you instrument your JVM-based application code without vendor lock-in. You can configure the [OpenTelemetry Micrometer Bridge](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/micrometer/micrometer-1.5/library) along with the OpenTelemetry SDK to forward Micrometer data to New Relic via OTLP.

Reviewer 2: @akristen I would remove this marketing-y intro sentence and replace it with stuff related to how New Relic can monitor it/what data we monitor/why its beneficial) (since we can assume they know what micrometer is by definition/they don't need high-level concept references)

> This doc walks you through procedures to forward Micrometer data with an example Spring app. These are generalized steps that you may want to update to fit your specific environment. 

Reviewer 2: ```suggestion
This doc walks you through procedures to forward Micrometer data with a sample Spring app. These are generalized steps that you may want to update to fit your specific environment. 
```
